boxtree/tree_build.py

```diff
@@ -37,6 +37,251 @@ class MaxLevelsExceeded(RuntimeError):
     pass
 
 
+# {{{ box-based tree
+
+def resized_array(arr, new_size):
+    """Return a resized copy of the array. The new_size is a scalar which is
+    applied to the last dimension.
+    """
+    old_size = arr.shape[-1]
+    prefix = (slice(None), ) * (arr.ndim - 1)
+    if old_size >= new_size:
+        return arr[prefix + (slice(new_size), )].copy()
+    else:
+        new_shape = list(arr.shape)
+        new_shape[-1] = new_size
+        new_arr = np.zeros(new_shape, arr.dtype)
+        new_arr[prefix + (slice(old_size), )] = arr
+        return new_arr
+
+
+def vec_of_signs(dim, i):
+    """The sign vector is obtained by converting i to a dim-bit binary.
+    """
+    # e.g. bin(10) = '0b1010'
+    binary_digits = [int(bd) for bd in bin(i)[2:]]
+    n = len(binary_digits)
+    assert n <= dim
+    return np.array([0]*(dim-n) + binary_digits) * 2 - 1
+
+
+class _TreeOfBoxes:
```

Comment:
I think it'd be useful to convey whether the arrays in here are expected to be numpy arrays, CL arrays, or something else entirely.